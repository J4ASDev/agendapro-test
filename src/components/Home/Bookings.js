import React from 'react';
import PropTypes from 'prop-types';
import { Droppable, Draggable } from 'react-beautiful-dnd';

const BookingsWrapper = ({ droppableId, bookings, onCancelBooking }) => (
  <Droppable droppableId={droppableId}>
    { provided => (
      <article>
        <div className='title'>
          <h4>{droppableId.slice(0, -1)}</h4> <span>{bookings.length}</span>
        </div>
        <div ref={provided.innerRef} className='bookingsContainer'>
          { bookings.map((item, index) => {
            const { client: { first_name, email }, start, end, price, id } = item;

            return (
              <Draggable key={id} draggableId={id} index={index}>
                { provided => (
                  <div
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    className='bookingItem'
                  >
                    <div>
                      <span>Name:</span> <p>{first_name}</p>
                    </div>
                    <div>
                      <span>Email:</span> <p>{email}</p>
                    </div>
                    <div>
                      <span>Start Date:</span> <p>{start.slice(0, 10)}</p>
                    </div>
                    <div>
                      <span>Start Time:</span> <p>{start.slice(11, 19)}</p>
                    </div>
                    <div>
                      <span>End Date:</span> <p>{end.slice(0, 10)}</p>
                    </div>
                    <div>
                      <span>End Time:</span> <p>{end.slice(11, 19)}</p>
                    </div>
                    <div>
                      <span>Price:</span> <p>{price}</p>
                    </div>
                    <button onClick={() => onCancelBooking(id, bookings, droppableId)}>
                      Cancel booking
                    </button>
                  </div>
                )}
              </Draggable>
            );
          })}
          {provided.placeholder}
        </div>
      </article>
    )}
  </Droppable>
);

BookingsWrapper.propTypes = {
  droppableId: PropTypes.string,
  bookings: PropTypes.array,
  onCancelBooking: PropTypes.func
};

export default BookingsWrapper;
