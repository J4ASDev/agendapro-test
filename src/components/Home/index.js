import React from 'react';
import PropTypes  from 'prop-types';
import { DragDropContext } from 'react-beautiful-dnd';
import BookingsWrapper from './Bookings';

const HomeLayout = ({
  reserved,
  confirmed,
  attends,
  canceled,
  notAttends,
  standby,
  pending,
  onDragEnd
}) => (
  <DragDropContext onDragEnd={onDragEnd}>
    <BookingsWrapper droppableId='reserved1' bookings={reserved} />
    <BookingsWrapper droppableId='confirmed2' bookings={confirmed} />
    <BookingsWrapper droppableId='attends3' bookings={attends} />
    <BookingsWrapper droppableId='canceled5' bookings={canceled} />
    <BookingsWrapper droppableId='notAttends6' bookings={notAttends} />
    <BookingsWrapper droppableId='standby7' bookings={standby} />
    <BookingsWrapper droppableId='pending8' bookings={pending} />
  </DragDropContext>
);

HomeLayout.propTypes = {
  reserved: PropTypes.array,
  confirmed: PropTypes.array,
  attends: PropTypes.array,
  canceled: PropTypes.array,
  notAttends: PropTypes.array,
  standby: PropTypes.array,
  pending: PropTypes.array
};

export default HomeLayout;
