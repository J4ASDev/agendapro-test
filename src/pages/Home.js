import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../actions/home';
import HomeLayout from '../components/Home';

class Home extends Component {
  idStatusLists = {
    reserved1: 'reserved',
    confirmed2: 'confirmed',
    attends3: 'attends',
    notAttends6: 'notAttends',
    standby7: 'standby',
    pending8: 'pending'
  };

  getList = id => this.props[this.idStatusLists[id]];

  onDragEnd = ({ source, destination, draggableId: dragBookingId }) => {
    const { reorderBookingsDnd, moveBookingsDnd } = this.props.actions;
    const sourceDropId = source.droppableId;
    const destDropId = destination.droppableId;
    const sourceDropList = this.getList(sourceDropId);
    const destDropList = this.getList(destDropId);

    // dropped outside the list
    if (!destination) return;

    if (sourceDropId === destDropId) {
      reorderBookingsDnd(
        sourceDropId,
        sourceDropList,
        source.index,
        destination.index
      );
    } else {
      const booking = sourceDropList.find(booking => booking.id === dragBookingId);
      const bookingDate = new Date(booking.start.slice(0, 10));
      const currentDate = new Date();

      if (bookingDate >= currentDate) {
        moveBookingsDnd( sourceDropList, destDropList, source, destination );
      } else {
        console.log('Given date is not greater than the current date');
      }
    }
  };

  onCancelBooking = (id, bookings, idList) => {
    this.props.actions.cancelBooking(id, bookings, idList);
  }

  onFilterDateByRange = event => {
    event.preventDefault();

    const { getBookings } = this.props.actions;

    const form = document.getElementById('form-filter-date');
    const data = new FormData(form);
    const rangeFrom = data.get('range-from');
    const rangeTo = data.get('range-to');

    if (rangeFrom && rangeTo) return getBookings(69205, rangeFrom, rangeTo);

    return getBookings(69205);
  }

  render() {
    const { reserved, confirmed, attends, notAttends, standby, pending } = this.props;
    return (
      <HomeLayout
        reserved={reserved}
        confirmed={confirmed}
        attends={attends}
        notAttends={notAttends}
        standby={standby}
        pending={pending}
        onFilterDateByRange={this.onFilterDateByRange}
        onCancelBooking={this.onCancelBooking}
        onDragEnd={this.onDragEnd}
      />
    );
  }

  componentDidMount() {
    this.props.actions.getBookings(69205);
  }
}

Home.propTypes = {
  reserved: PropTypes.array,
  confirmed: PropTypes.array,
  attends: PropTypes.array,
  notAttends: PropTypes.array,
  standby: PropTypes.array,
  pending: PropTypes.array,
  actions: PropTypes.object,
  getBookings: PropTypes.func,
  cancelBooking: PropTypes.func,
  reorderBookingsDnd: PropTypes.func,
  moveBookingsDnd: PropTypes.func
};

const mapStateToProps = ({ home }) => ({
  reserved: home.reserved,
  confirmed: home.confirmed,
  attends: home.attends,
  notAttends: home.notAttends,
  standby: home.standby,
  pending: home.pending
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
