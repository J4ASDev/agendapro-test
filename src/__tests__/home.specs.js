import React from 'react';
import { shallow } from 'enzyme';
import Home from '../components/Home';

describe('#Home', () => {
  let props;
  let wrapper;

  beforeEach(()=> {
    props = {
      onCancelBooking: () => {},
      onDragEnd: () => {},
      onFilterDateByRange: () => {},
      reserved: [],
      confirmed: [],
      attends: [],
      notAttends: [],
      standby: [],
      pending: []
    };

    wrapper = shallow(<Home {...props} />);
  });

  it('renders without crashing', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
