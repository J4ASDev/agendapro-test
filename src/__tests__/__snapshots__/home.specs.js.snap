// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Home renders without crashing 1`] = `
"<div className=\\"wrapper\\">
  <div>
    <h1>
      AgendaPro Test - Bookings - Jonathan Araujo
    </h1>
    <form id=\\"form-filter-date\\" onSubmit={[Function: onFilterDateByRange]}>
      <label>
         From: 
        <input type=\\"date\\" name=\\"range-from\\" />
         
      </label>
      <label>
         To: 
        <input type=\\"date\\" name=\\"range-to\\" />
         
      </label>
      <button>
        Filter
      </button>
    </form>
  </div>
  <div className=\\"wrapper-bookings\\">
    <DragDropContext onDragEnd={[Function: onDragEnd]}>
      <BookingsWrapper droppableId=\\"reserved1\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
      <BookingsWrapper droppableId=\\"confirmed2\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
      <BookingsWrapper droppableId=\\"attends3\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
      <BookingsWrapper droppableId=\\"notAttends6\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
      <BookingsWrapper droppableId=\\"standby7\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
      <BookingsWrapper droppableId=\\"pending8\\" bookings={{...}} onCancelBooking={[Function: onCancelBooking]} />
    </DragDropContext>
  </div>
</div>"
`;
